secret-store/secret_store.py

```diff
@@ -0,0 +1,145 @@
+
+
+@dataclass(frozen=True)
+class Table:
+    name: str
+    columns: list[str]
+
+
+SECRETS_TABLE = Table("secrets", ["user", "password"])
+CONFIG_ITEMS_TABLE = Table("config_items", ["item"])
+
+
+@dataclass(frozen=True)
+class Credentials:
+    user: str
+    password: str
+
+
+class Secrets:
+    def __init__(self, db_file: str, master_password: str) -> None:
+        self.db_file = db_file
+        self.master_password = master_password
+        self._con = None
+        self._cur = None
+
+    def close(self) -> None:
+        if self._cur is not None:
+            self._cur.close()
+            self._cur = None
+        if self._con is not None:
+            self._con.close()
+            self._con = None
+
+    def connection(self) -> sqlcipher.Connection:
+        if self._con is None:
+            if not os.path.exists(self.db_file):
+                _logger.info(f"Creating file {self.db_file}")
+            self._con = sqlcipher.connect(self.db_file)
+            self._use_master_password()
+            self.create_tables()
+        return self._con
+
+    def _use_master_password(self) -> None:
+        """
+        If database is unencrypted then this method encrypts it.
+        If database is already encrypted then this method enables to access the data.
+        """
+        if self.master_password is not None:
+            sanitized = self.master_password.replace("'", "\\'")
+            self.cursor().execute(f"PRAGMA key = '{sanitized}'")
+
+    def cursor(self) -> sqlcipher.Cursor:
+        if self._cur is None:
+            self._cur = self.connection().cursor()
+        return self._cur
+
+    def has_table(self, name: str) -> bool:
+        res = self.cursor().execute("SELECT * FROM sqlite_master where name = ?", [name])
+        return True if res and res.fetchone() else False
+
+    # key represents a system, service, or application
+    def create_table(self, table: Table) -> None:
+        if self.has_table(table.name):
+            return
+        _logger.info(f'Creating table "{table.name}".')
+        columns = " ,".join(table.columns)
+        self.cursor().execute(f"CREATE TABLE {table.name} (key, {columns})")
+        self.connection().commit()
+
+    def create_tables(self) -> None:
+        for table in (SECRETS_TABLE, CONFIG_ITEMS_TABLE):
+            self.create_table(table)
+
+    def _save_data(self, table: Table, key: str, data: list[str]) -> "Secrets":
+        cur = self.cursor()
+        res = cur.execute(f"SELECT * FROM {table.name} WHERE key=?", [key])
+        if res and res.fetchone():
+            columns = ", ".join(f"{c}=?" for c in table.columns)
+            cur.execute(
+                f"UPDATE {table.name} SET {columns} WHERE key=?",
+                data + [key])
+        else:
+            columns = ", ".join("?" for c in table.columns)
+            cur.execute(
+                f"INSERT INTO {table.name} VALUES (?, {columns})",
```

Comment:
This assumes a defined column order. Better specify column names explicitly if possible

Reply:
Will be fixed in next push